### DISCUSSION TITLE
Slowness of spack

### DISCUSSION THREAD

**Discussion by 1a2a34v** [spack/spack#32234] [General]
I really like the idea of spack and have been using it with great success for my set of workstations to manage my software packages and their dependencies. However, I couldn't help but notice the very slow speed at which packages are installed compared to other package managers. For example, when using pip to install scipy it barely takes 30 seconds, while spack takes a lot longer to complete. I realize that spack is doing a lot more in the background than pipy is, but I was still wondering what the reason for this comparatively very slow installation speed is.

**Comment by sethrj** [comment]
Spack installs everything from source (requiring expensive configure, compile, link, and install phases) whereas most other packages simply unpack a precompiled set of libraries and applications to the destination. The reason is primarily because Spack offers far more configurable options than most other package managers: arbitrary combinations of different software versions, different options for each software, and different platforms. Homebrew, Apt, etc. typically support only a single version with a single preconfigured set of options. One of the primary drivers for Spack is also HPC platforms, where no other package manager can account for the highly customized system architecture.

**Reply by tgamblin** [reply]
This is all true, but there are a lot of steps we can take to make things faster than they are currently, and we're working on them.  I think you can expect Spack to get consistently faster as we go forward.  Some low hanging fruit:

1. Our binary installation is slow because we dynamically search for potential relocations at install time.  That can be cached, e.g. as in https://github.com/spack/spack/pull/30957
2. Version comparison can be slow, but there are ways to speed that up by leaning on tuples more internally: https://github.com/spack/spack/pull/30544.  This will speed up concretization as there is a lot of version comparison in the setup step.
3. The setup step of concretization is *mostly* consumed by loading `package.py` files for all possible dependencies of what you asked for (which can be like 400-500 files or more).  Python loading is slow.  I believe we can cache the metadata from packages and speed this part up a *lot*, as in many cases it accounts for most of the solve time.  I'll be working on that pretty soon here.
4. We can't make source builds fundamentally faster, but we can make sure that more binaries are available -- this should also help with installation speed.  We continue to expand the [binary cache](https://spack.io/spack-binary-packages/).

Those are a few speedups currently in the queue.
